File: First-Edition/BookSamples/rosetta-stone/scala3/rsc2p1p3_traits_type_classes/src/main/scala/assets/GoL.scala
import Automaton.*

object GoL:

  case class GoLWorld(board: Board)

  given golAutomaton: IAutomaton[GoLWorld] with
    extension (world: GoLWorld) def step: GoLWorld =
      val newBoard = world.board.indices.map { x =>
        world.board(x).indices.map { y =>
          nextState(world.board, x, y)
        }.toVector
      }.toVector
      GoLWorld(newBoard)

    extension (board: Board) def wrap: GoLWorld = GoLWorld(board)
    extension (world: GoLWorld) def unwrap: Board = world.board

  def nextState(board: Board, x: Int, y: Int): Cell = {
    val neighbors = for {
      i <- -1 to 1
      j <- -1 to 1
      if i != 0 || j != 0
      if board.isDefinedAt(x + i) && board(x + i).isDefinedAt(y + j)
    } yield board(x + i)(y + j)

    val aliveNeighbors = neighbors.count(_ == Alive)

    board(x)(y) match {
      case Alive if aliveNeighbors < 2 || aliveNeighbors > 3 => Dead
      case Alive => Alive
      case Dead if aliveNeighbors == 3 => Alive
      case Dead => Dead
    }
  }

File: First-Edition/BookSamples/rosetta-stone/rust/rsc1p1p3_traits_type_classes/src/assets/gol.rs

use crate::automaton;

use automaton::Board;
use automaton::Cell;
use automaton::Automaton;

#[derive(Clone)]
pub struct GoL(pub Board);

fn gol_step(board: &Board) -> Board {
  let mut new_board = board.clone();

  for (i, row) in board.iter().enumerate() {
    for (j, cell) in row.iter().enumerate() {
      let nh = automaton::get_neighbor_alive(i, j, &board);

      new_board[i][j] = match (cell, nh) {
        (Cell::Alive, 2) => Cell::Alive,
        (Cell::Alive, 3) => Cell::Alive,
        (Cell::Dead,  3) => Cell::Alive,
        _ => Cell::Dead,
      }
    }
  }

  return new_board;
}

impl Automaton for GoL {
  fn step(&self) -> Self {
    return GoL(gol_step(&self.0));
  }
}
